src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1121,10 +1166,86 @@ private static ImmutableArray<StateForCase> RemoveEvaluation(ImmutableArray<Stat
             falseTestPermitsTrueOther = true;
             trueTestImpliesTrueOther = false;
             falseTestImpliesTrueOther = false;
+            precondition = Tests.True.Instance;
+            sideeffect = Tests.True.Instance;
 
             // if the tests are for unrelated things, there is no implication from one to the other
-            if (!test.Input.Equals(other.Input))
-                return;
+            BoundDagTemp s1Input = test.Input;
+            BoundDagTemp s2Input = other.Input;
+            if (!s1Input.Equals(s2Input))
```

Comment:
nit: not sure if it'll read better, but worth considering to reduce depth of this method: we could remove this `if` and use a `while (!s1Input.Equals(s2Input))` loop below instead of a `do/while`. Then we can lazily initialize `preconditions` when we need it. If `preconditions` is initialized at the end of the `while` loop, then we set `precondition`/`sideeffect`

Reply:
Do we have a perf concern for the common case of empty precondition?
Otherwise the current flow is needed per https://github.com/dotnet/roslyn/pull/53891#discussion_r691470124

Reply:
Just to clarify, I was not proposing to change the logic, only to refactor/flatten it. Something like:

```
ArrayBuilder<Tests> preconditions = null;
while (!s1Input.Equals(s2Input))
{
    switch (s1Input.Source, s2Input.Source)
    {
        ... do `preconditions ??= ArrayBuilder<Tests>.GetInstance();` when we need it ...
        ... code in switch remains as-is ...
    }

    // tests are unrelated
    preconditions?.Free();
    return;
}

if (preconditions is not null)
{
                precondition = Tests.AndSequence.Create(preconditions);
                // At this point, we have determined that two non-identical inputs refer to the same element.
                // We represent this correspondence with an assignment node in order to merge the remaining values.
                sideeffect = new Tests.One(new BoundDagAssignmentEvaluation(syntax, target: other.Input, input: test.Input));
                // PROTOTYPE(list-patterns): Do we still need to add this node if there's no preconditions?
}

switch (test)
...
```

Reply:
I understood what you were suggesting. This changes the logic as we no longer set sideeffect if there's no precondition with no way to know. That's what the prototype comment is tracking (because we don't have a test case where this matters and that hasn't been proved otherwise.)